File: leetcode_312.cpp
class Solution {
public:
  string countAndSay(int n) {
    string s("1");
    while (--n)
      s = getNext(s);
    return s;
  }

  string getNext(const string &s) {
    stringstream ss;
    for (auto i = s.begin(); i !=  s.end; i) {
      auto j = find_if(i, s.end(), bind1st(not_equal_to<char>(), *i));
      ss << distance(i, j) << *i;
      i = j;
    }
    return ss.str();
  }
}

File: leetcode_312.py
class Solution:
    # @return a string
    def countAndSay(self, n):
        oldStr = '1'
	for i in range(n, 0, -1):
	  newStr = ''
	  j = 0
	  while j < len(oldStr):
	    count = 1
	    while j + 1 < len(oldStr) and oldStr[j] == oldStr[j + 1]:
	      count += 1
	      j += 1
	    newStr += str(count) + oldStr[j]
	    j += 1    
	  oldStr = newStr

	return oldStr  


s = Solution()
print s.countAndSay(3)
